www/js/lib/zimfile.js

```diff
@@ -48,89 +47,98 @@ define(['xzdec_wrapper', 'zstddec_wrapper', 'util', 'utf8', 'q', 'zimDirEntry'],
+     * @param {Integer} offset The absolute offset from the start of the ZIM file or file set at which to start reading
+     * @param {Integer} size The number of bytes to read
+     * @returns {Promise<Uint8Array>} A Promise for a Uint8Array containing the requested data
+     */
+    ZIMFile.prototype._readSlice = function(offset, size) {
+        return FileCache.read(this, offset, offset + size);
+    };
+
+    /**
+     * Read a slice from a set of one or more ZIM files constituting a single archive, and concatenate the data parts
+     * @param {Integer} begin The absolute byte offset from which to start reading 
+     * @param {Integer} end The absolute byte offset where reading should stop (the end byte is not read)
+     * @returns {Promise<Uint8Array>} A Promise for a Uint8Array containing the concatenated data 
      */
-    ZIMFile.prototype._readSlice = function(offset, size)
-    {
+    ZIMFile.prototype._readSplitSlice = function (begin, end) {
+        var file = this;
         var readRequests = [];
         var currentOffset = 0;
-        for (var i = 0; i < this._files.length; currentOffset += this._files[i].size, ++i) {
-            var currentSize = this._files[i].size;
-            if (offset < currentOffset + currentSize && currentOffset < offset + size) {
-                var readStart = Math.max(0, offset - currentOffset);
-                var readSize = Math.min(currentSize, offset + size - currentOffset - readStart);
-                readRequests.push(util.readFileSlice(this._files[i], readStart, readSize));
+        for (var i = 0; i < file._files.length; currentOffset += file._files[i].size, ++i) {
+            var currentSize = file._files[i].size;
+            if (begin < currentOffset + currentSize && currentOffset < end) {
+                // DEV: Math.max is used below because we could be reading the last part of a blob split across two files,
+                // in which case (begin - currentOffset) could be negative!
+                var readStart = Math.max(0, begin - currentOffset);
+                var readEnd = Math.min(currentSize, end - currentOffset);
+                readRequests.push(util.readFileSlice(file._files[i], readStart, readEnd));
             }
         }
-        if (readRequests.length == 0) {
+        if (readRequests.length === 0) {
             return Q(new Uint8Array(0).buffer);
-        } else if (readRequests.length == 1) {
+        } else if (readRequests.length === 1) {
             return readRequests[0];
         } else {
             // Wait until all are resolved and concatenate.
             console.log("CONCAT");
-            return Q.all(readRequests).then(function(arrays) {
-                var concatenated = new Uint8Array(size);
-                var sizeSum = 0;
-                for (var i = 0; i < arrays.length; ++i) {
-                    concatenated.set(new Uint8Array(arrays[i]), sizeSum);
-                    sizeSum += arrays[i].byteLength;
-                }
```

Comment:
Consider reverting code below to above (largely). Although code below works fine, it was part of an early attempt to correct the split ZIM archive issue. Original code would be marginally faster and the first arrays.forEach is no longer needed.

Reply:
Done.